rfcs/0010-mempool-dos-mitigation.md

```diff
@@ -0,0 +1,302 @@
+nonce. Users can e.g. cancel payments by replacing them with a no-op transaction
+(send themselves $0) or resend them with a higher fee if they're processing too
+slow.
+
+We want to have this feature, but it allows an attacker to make proposers
+process transactions which won't eventually get mined (the ones that are
+replaced), which violates our first guiding principle. So we require an added
+fee, `mempool_replace_fee`, representing the cost to the network to process and
+store a pending transaction, and add a field to transaction payloads counting
+the number of replacements that occurred for a given nonce. A transaction's
+total fee is its original transaction fee plus the replacement count times
+`mempool_replace_fee`. If a node receives a new transaction with the same nonce
+and sender account as one in the mempool, it looks at the replacement count and
+checks if it is different from the replacement counts of all the transactions
+from that account/nonce pair it's seen so far. If it is, and the fee is higher
+it will include that transaction in the next block it mines. If the
+account/nonce/replacement count triple has been seen before then we know the
+payment sender is attempting to cheat and has sent two or more transactions
+without paying the correct replacement fee.
+
+In that case, we punish the *sender*. It's not possible to solve this problem by
+blacklisting peers, because peers may not be aware that a transaction is an
+illegal duplicate. Imagine Mallory connects to two proposers, Alice and Bob. She
+sends two different transactions, both with nonce n, to Alice and Bob
+simultaneously. Alice and Bob gossip these new transactions to each other, and
+both think their peer is sending them an invalid transaction. If we used peer
+blacklisting here, Mallory could cause Alice and Bob to blacklist each other,
+even though they were both acting honestly.
+
+To punish the sender, we require senders to "deposit" an amount equal to
+twice the transaction fee when sending any transaction. We also add a new type
+of transaction, a fraud proof, that proves a sender sent two or more
+transactions with the same account/nonce/replacement count triple, by supplying
+the two transactions. If one of these transactions is included in the
+blockchain, the miner gets the deposit. If no such transaction is included after
+`fraud_proof_window`, the deposit is returned. The deposit is twice the
+transaction fee because if it's forfeited we need to eventually make a SNARK
+that checks two transactions. We rely again on the fact that SNARK proving is
+much more expensive than transaction processing. This scheme has the unfortunate
+downside that we need to track when deposits are returned to keep balances up to
+date. Not sure how much overhead that adds.
+
+#### Multiple pending transactions from the same account
+
+Since payments aren't instant, users may want to queue multiple outgoing
```

Comment:
Are there alternatives to re-use of nonces that would rid us of some of these issues?